File: sdk/typescript/src/runtime/response.ts
// @waeasi/sdk — Response
//
// Fetch-API compatible response.  The user constructs it with any of
// the standard BodyInit forms; the runtime harness later flushes it to
// wasi:http via `writeToOutparam(...)`.
//
// We deliberately keep status-text minimal: HTTP/2 dropped reason
// phrases entirely, and the kernel router maps numeric status to the
// canonical IANA reason for HTTP/1.1 wire formatting.

import { Body, type BodyInit } from "./body.js";
import { Headers, type HeadersInit } from "./headers.js";
import type { ResponseOutparam } from "../wasi/http.js";

export interface ResponseInit {
    status?: number;
    statusText?: string;
    headers?: HeadersInit;
}

export type JsonValue =
    | string | number | boolean | null
    | { [k: string]: JsonValue }
    | JsonValue[];

export class Response {
    readonly status: number;
    readonly statusText: string;
    readonly headers: Headers;
    private readonly _body: Body;

    constructor(body?: BodyInit, init?: ResponseInit) {
        const status = init?.status ?? 200;
        if (status < 100 || status > 599) {
            throw new RangeError(`status out of range: ${status}`);
        }
        this.status = status;
        this.statusText = init?.statusText ?? "";
        this.headers = new Headers(init?.headers);
        this._body = new Body(body);
    }

    get ok(): boolean       { return this.status >= 200 && this.status < 300; }
    get bodyUsed(): boolean { return this._body.bodyUsed; }

    arrayBuffer(): Promise<ArrayBuffer> { return this._body.arrayBuffer(); }
    bytes(): Promise<Uint8Array>        { return this._body.bytes(); }
    text(): Promise<string>             { return this._body.text(); }
    json<T = unknown>(): Promise<T>     { return this._body.json<T>(); }

    /** Static helper — `Response.json({...})` with proper Content-Type. */
    static json(value: JsonValue, init?: ResponseInit): Response {
        const headers = new Headers(init?.headers);
        if (!headers.has("content-type")) {
            headers.set("content-type", "application/json; charset=utf-8");
        }
        return new Response(JSON.stringify(value), {
            status: init?.status,
            statusText: init?.statusText,
            headers,
        });
    }

    /** Static helper — 3xx redirect. */
    static redirect(location: string, status: 301 | 302 | 303 | 307 | 308 = 302): Response {
        return new Response(null, {
            status,
            headers: { location },
        });
    }

    /** Static helper — error response with text/plain body. */
    static error(status: number, message?: string): Response {
        return new Response(message ?? "", {
            status,
            headers: { "content-type": "text/plain; charset=utf-8" },
        });
    }

    /**
     * Hostside finalizer — write status + headers + body into a wasi:http
     * response-outparam.  Auto-injects Content-Length when possible.
     */
    async writeToOutparam(out: ResponseOutparam): Promise<void> {
        // Best-effort Content-Length: we already have the bytes if the body
        // was constructed from string/Uint8Array.  The harness will skip
        // the header for streaming bodies (transfer-encoding: chunked).
        if (!this.headers.has("content-length")
            && !this.headers.has("transfer-encoding")
        ) {
            const cl = await this.tryComputeContentLength();
            if (cl !== null) this.headers.set("content-length", String(cl));
        }
        const tx = out.startResponse(this.status, this.headers.toWasi());
        try {
            await this._body.pipeToWasi(tx);
        } finally {
            tx.close();
            out.finish();
        }
    }

    private async tryComputeContentLength(): Promise<number | null> {
        // Only fast-path for already-materialised bodies.
        const internal = (this._body as unknown as { source: { kind: string; data?: Uint8Array } }).source;
        if (internal.kind === "empty") return 0;
        if (internal.kind === "bytes" && internal.data) return internal.data.byteLength;
        return null;
    }
}

File: sdk/python/src/waeasi/runtime/response.py
"""waeasi.runtime.response — Fetch-style Response.

Mirrors :class:`Request` but on the egress side.  `Response.json(...)`
and `Response.error(...)` static helpers cover the common cases.
"""

from __future__ import annotations

import json as _json
from typing import Any, Optional

from ..wasi.http import ResponseOutparam
from .body import Body, BodyInit
from .headers import Headers, HeadersInit


class Response:
    __slots__ = ("status", "status_text", "headers", "_body")

    def __init__(
        self,
        body: BodyInit = None,
        status: int = 200,
        status_text: str = "",
        headers: HeadersInit = None,
    ) -> None:
        if not (100 <= status <= 599):
            raise ValueError(f"status out of range: {status}")
        self.status = status
        self.status_text = status_text
        self.headers = Headers(headers)
        self._body = Body(body)

    @property
    def ok(self) -> bool:
        return 200 <= self.status < 300

    @property
    def body_used(self) -> bool:
        return self._body.body_used

    async def bytes(self) -> bytes:
        return await self._body.bytes_()

    async def text(self) -> str:
        return await self._body.text()

    async def json(self) -> Any:
        return await self._body.json()

    @classmethod
    def json_response(
        cls,
        value: Any,
        status: int = 200,
        headers: HeadersInit = None,
    ) -> "Response":
        h = Headers(headers)
        if not h.has("content-type"):
            h.set("content-type", "application/json; charset=utf-8")
        return cls(_json.dumps(value, ensure_ascii=False), status=status, headers=h)

    @classmethod
    def redirect(cls, location: str, status: int = 302) -> "Response":
        if status not in (301, 302, 303, 307, 308):
            raise ValueError(f"redirect status not allowed: {status}")
        return cls(None, status=status, headers={"location": location})

    @classmethod
    def error(cls, status: int, message: Optional[str] = None) -> "Response":
        return cls(
            message or "",
            status=status,
            headers={"content-type": "text/plain; charset=utf-8"},
        )

    async def write_to_outparam(self, out: ResponseOutparam) -> None:
        if (not self.headers.has("content-length")
            and not self.headers.has("transfer-encoding")):
            cl = await self._try_content_length()
            if cl is not None:
                self.headers.set("content-length", str(cl))
        tx = out.start_response(self.status, self.headers.to_wasi())
        try:
            await self._body.pipe_to_wasi(tx)
        finally:
            tx.close()
            out.finish()

    async def _try_content_length(self) -> Optional[int]:
        # Only fast path for bodies we already have in memory.
        if self._body._kind == "empty":  # type: ignore[attr-defined]
            return 0
        if self._body._kind == "bytes":  # type: ignore[attr-defined]
            return len(self._body._data)  # type: ignore[attr-defined]
        return None
